keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlResourceRepository.kt

```diff
@@ -239,21 +239,35 @@ open class SqlResourceRepository(
     jooq.transaction { config ->
       val txn = DSL.using(config)
 
+      val resourceUid = txn.select(RESOURCE.UID).from(RESOURCE).where(RESOURCE.ID.eq(event.id)).fetchOne(RESOURCE.UID)
+
       if (event.ignoreRepeatedInHistory) {
         val previousEvent = txn
           .select(EVENT.JSON)
-          .from(EVENT, RESOURCE)
+          .from(EVENT)
           .where(EVENT.SCOPE.eq(Scope.RESOURCE.name))
-          .and(RESOURCE.ID.eq(event.id))
-          .and(RESOURCE.UID.eq(EVENT.UID))
+          .and(EVENT.UID.eq(resourceUid))
           .orderBy(EVENT.TIMESTAMP.desc())
           .limit(1)
           .fetchOne()
           ?.let { (json) ->
             objectMapper.readValue<ResourceEvent>(json)
           }
 
-        if (event.javaClass == previousEvent?.javaClass) return@transaction
+        // If an equivalent event is already the latest in history, just update the timestamp
+        if (event.javaClass == previousEvent?.javaClass) {
+          txn
+            .update(EVENT)
+            .set(EVENT.TIMESTAMP, event.timestamp.atZone(clock.zone).toLocalDateTime())
+            // we udpate the JSON as well because it also contains a timestamp
+            .set(EVENT.JSON, objectMapper.writeValueAsString(event))
+            .where(EVENT.SCOPE.eq(Scope.RESOURCE.name))
+            .and(EVENT.UID.eq(resourceUid))
+            .orderBy(EVENT.TIMESTAMP.desc())
+            .limit(1)
```

Comment:
I assume this ordering and limiting is so you grab the most recent event? I think it would be more efficient to just match on the timestamp of the `previousEvent`, something like `.and(EVENT.TIMESTAMP.eq(previousEvent.timestamp)`? Or, we might have to also fetch the timestamp from the previous event, but I think that's going to be more efficient than fetching all of them here and then ordering them again

Reply:
Good idea. We should really add a proper PK to this table (I believe it was one of the suggestions we got from DET Platform). Should I also add the event type to the query to avoid the risk of two resource events emitted at the “same time” (ResourceDeltaResolved followed by ResourceValid?), or would those happen in separate check iterations?

Reply:
Those would happen on separate check iterations so I think you should be good with just time!